tsweb/tsweb.go

```diff
@@ -447,6 +449,55 @@ func VarzHandler(w http.ResponseWriter, r *http.Request) {
 	varz.Handler(w, r)
 }
 
+// CleanRedirectURL ensures that urlStr is a valid redirect URL to the
+// current server, or one of allowedHosts. Returns the cleaned URL or
+// a validation error.
+func CleanRedirectURL(urlStr string, allowedHosts []string) (*url.URL, error) {
+	// In some places, we unfortunately query-escape the redirect URL
+	// too many times, and end up needing to redirect to a URL that's
+	// still escaped by one level. Try to unescape the input.
+	unescaped, err := url.QueryUnescape(urlStr)
+	if err == nil && unescaped != urlStr {
+		urlStr = unescaped
+	}
+
+	// Go's URL parser and browser URL parsers disagree on the meaning
+	// of malformed HTTP URLs. Given the input https:/evil.com, Go
+	// parses it as hostname="", path="/evil.com". Browsers parse it
+	// as hostname="evil.com", path="". This means that, using
+	// malformed URLs, an attacker could trick us into approving of a
+	// "local" redirect that in fact sends people elsewhere.
+	//
+	// This very blunt check enforces that we'll only process
+	// redirects that are definitely well-formed URLs.
+	//
+	// Note that the check for just / also allows URLs of the form
+	// "//foo.com/bar", which are scheme-relative redirects. These
+	// must be handled with care below when determining whether a
+	// redirect is relative to the current host. Notably,
+	// url.URL.IsAbs reports // URLs as relative, whereas we want to
+	// treat them as absolute redirects and verify the target host.
+	if !strings.HasPrefix(urlStr, "/") && !strings.HasPrefix(urlStr, "http://") && !strings.HasPrefix(urlStr, "https://") {
+		return nil, fmt.Errorf("invalid redirect URL %q", urlStr)
+	}
+
+	url, err := url.Parse(urlStr)
+	if err != nil {
+		return nil, fmt.Errorf("invalid redirect URL %q", urlStr)
+	}
+	// Redirects to self are always allowed. A self redirect must
+	// start with url.Path, all prior URL sections must be empty.
+	isSelfRedirect := url.Scheme == "" && url.Opaque == "" && url.User == nil && url.Host == ""
+	if isSelfRedirect {
+		return url, nil
+	}
+	if !slices.Contains(allowedHosts, url.Hostname()) {
```

Comment:
Could get a false negative if `Hostname()` is not normalized to whatever case `allowedHosts` has. Probably want to normalize case before testing, to avoid surprises.

Reply:
Hm, again we've not needed to handle people playing games with case in hostnames so far, and this conservatively rejects such games. I'm worried that if we get into the business of case folding, we also have to worry about unicode case, punycode representations and so forth, rather than a straight byte-wise comparison. Do you think the added complexity is worth it, given that we don't seem to need it so far?

Reply:
With the potential false negative, I'm imagining someone writing code like `allowed := []string{"tailscale.com", "login.tailscale.com"}`, and then another person writing some React for the front end and saying something like `redirect("https://Tailscale.com/...")`, and then we spend 3 hours being confused why our allowed domain name is not being allowed, until someone finally notices the capital T.

I'm not worried about e.g. tailscαle.com (Greek alpha) or punycode, because those are just plain different domains that would be true negatives (i.e. not in `allowed`).

Reply:
Per discussion, updated to do case-folding comparison on the hostnames.